### DISCUSSION TITLE
Creating a custom DFN model with custom parameters

### DISCUSSION THREAD

**Discussion by vishankkumar** [pybamm-team/PyBaMM#1468] [Q&A] [answered]
Hi,

I have a simple question about how to create a custom DFN model similar to what is explained in the `custom_model.py` script but with reduced set of parameters. For instance if I use `no_sei` model, how can I avoid providing sei parameters as by default it takes default sei parameters?

Is there a way to add parameters as well one by one like for cathode, anode, separator and sei/no sei?

Secondly, is there a way to extract all the input parameters as a csv file, that can be read as an input to create a new model? I tried `extract_csv` method but it does not work as it does not print in the same manner as `parameters.csv`.

Thanks.

**Comment by valentinsulzer** [comment, marked as answer]
Look out for #1458 which when it is done should answer all your questions (@Saransh-cpp )

The model only needs the parameters that are actually in the model. So if you use the no_sei model and don't give any SEI parameters, it shouldn't error.

Parameters are defined separately for each part (though we call it negative electrode instead of anode). You can [pick and choose different components](https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/parameters/parameter_sets.py). Also, once you have created the `ParameterValues` object, you can update any individual parameters using `.update`

Not sure what you mean for the last one. You can save the parameters dictionary to a csv using inbuilt python functions, but it won't include functions. Why do you need to go via the csv? Can you give an example?

**Reply by vishankkumar** [reply]
Thanks @tinosulzer for your quick response. The thread was useful to answer some questions. 

The pick and choose option is only valid for the parameters registered in the `parameter set`. Although it is not very difficult to systematically add your parameter set there using built-in functions, my question was more related to if you can update part of the parameter set like for `negative electrode`. For testing purpose, I am using the following method:

```
# load model
model = lithium_ion.BasicDFN()
default_parameters = model.default_parameter_values

# Read my anode parameters from csv file
anode_param = pybamm.ParameterValues("parameters/lithium-ion/anodes/my_anode/parameters.csv")

# Update parameter values
orig_values = ParameterValues(chemistry=pybamm.parameter_sets.Chen2020)
orig_values.update(anode_param) 
```

This gives an attribute error: `AttributeError: 'int' object has no attribute 'lower'`. Will you please tell me what I am doing wrong here and how can I achieve this?

For the second question, I am trying to find a way where I can enter all the parameter values from a csv file. For testing again, I am trying to do something like this:

```
# load model
model = lithium_ion.BasicDFN()
default_parameters = model.default_parameter_values

# Extracting parameter values
values = ParameterValues(chemistry=pybamm.parameter_sets.Chen2020)

# Exporting values to  csv file
values.export_csv(filename="export_param.csv")

# Reading values from this csv file
new_values = pybamm.ParameterValues("export.csv")
```

And this gives a KeyError: `KeyError: 'Name [units]'`. Perhaps the exported csv is not the correct format to be read as input variables. Can you suggest any other option to read all the parameters from one single csv file, if possible?

Thank you for your response in advance.

**Reply by rtimms** [reply]
For the first, you can use 
```python3
model = pybamm.lithium_ion.BasicDFN()
values = model.default_parameter_values
anode_param = pybamm.ParameterValues("pybamm/input/parameters/lithium_ion/negative_electrodes/graphite_Chen2020/parameters.csv")
values.update(anode_param._dict_items) 
```
as a workaround. `update` expects a dict and something breaks when you pass a `ParameterValues` instance directly. Passing `anode_param._dict_items` gets you around the problem. I'll open an issue for this too. 

The second these fails because pybamm expects the columns of the csv to be labelled "Name [units]" and "Values". If you add these manually your example works. I'll make an issue so that `export_csv` adds the headers for you.

**Reply by Saransh-cpp** [reply]
@vishankkumar now that #1471 is merged, could you please try running your script again after pulling the latest code from the `develop` branch? Let us know if you face any issues.

**Comment by vishankkumar** [comment]
Hi Saransh,

Thank you for your email and working on the issue. However, I tried to run
some example scripts and it throws an error message during solving. The
error comes in the "casadi.py" file with the following error message:


AttributeError: 'DM' object has no attribute '__isnan__'


Let me know if I am missing something. This error I got even when trying
the simplest example script of "DFN.py"

Thanks for your response in advance.

Regards,
Vishank

**Reply by Saransh-cpp** [reply]
Could you paste the whole error here? I tried running the script on my side and everything worked fine.

**Comment by vishankkumar** [comment]
Sure,

Here is the full error I got:

  File
"/home/vishank-hp/miniconda3/lib/python3.7/site-packages/casadi-3.5.2b1-py3.7-linux-x86_64.egg/casadi/casadi.py",
line 83, in _swig_getattr
    raise AttributeError("'%s' object has no attribute '%s'" %
(class_type.__name__, name))
AttributeError: 'DM' object has no attribute '__isnan__'

Let me know if you need some other details to resolve this issue. Just FYI,
I pulled the latest version of "develop" branch and recompiled by using
"python setup.py develop"

Regards,
Vishank

**Reply by valentinsulzer** [reply]
If you do `pip install -e .` you don't need to recompile every time you pull changes from develop
That looks like an old error. What does `pip show pybamm` give?

**Comment by vishankkumar** [comment]
Hello Valtentin,

Thank you for your reply, I did pip install -e ., but still the same error. pip
show pybamm gave the following message:

Name: pybamm
Version: 0.4.0
Summary: Python Battery Mathematical Modelling.
Home-page: https://github.com/pybamm-team/PyBaMM
Author: None
Author-email: None
License: UNKNOWN
Location: /home/vishank-hp/github/PyBaMM
Requires: numpy, scipy, pandas, anytree, autograd, scikit-fem, casadi, jax,
jaxlib, jupyter, pybtex, matplotlib
Required-by:

Let me know if I am making a mistake.

Thanks and regards,
Vishank

**Reply by vishankkumar** [reply]
Hey,

I managed to solve the problem. Apparently the issue was with casadi solver, I uninstalled and reinstalled it and now everything works fine. Just to inform if anyone else facing the same issue

Regards,
Vishank
